File: CS/WinSolution.Module.Win/PromptClosingLastMainWindowController.cs
using System;
using System.Windows.Forms;
using DevExpress.ExpressApp;
using System.ComponentModel;
using DevExpress.XtraEditors;
using DevExpress.ExpressApp.Win;
using DevExpress.ExpressApp.Win.SystemModule;

namespace WinSolution.Module.Win {
    public class PromptClosingLastMainWindowController : WindowController {
        private bool isEditModelExecuting = false;
        EditModelController editModelController;
        protected override void OnActivated() {
            base.OnActivated();
            editModelController = Window.GetController<EditModelController>();
            if(editModelController != null) {
                editModelController.EditModelAction.Executing += new CancelEventHandler(EditModelAction_Executing);
                editModelController.EditModelAction.ExecuteCompleted += new EventHandler<DevExpress.ExpressApp.Actions.ActionBaseEventArgs>(EditModelAction_ExecuteCompleted);
            }
        }
        private void EditModelAction_ExecuteCompleted(object sender, DevExpress.ExpressApp.Actions.ActionBaseEventArgs e) {
            isEditModelExecuting = false;
        }
        private void EditModelAction_Executing(object sender, CancelEventArgs e) {
            isEditModelExecuting = true;
        }
        protected override void OnWindowChanging(Window window) {
            base.OnWindowChanging(window);
            window.TemplateChanged += OnWindowTemplateChanged;
        }
        private void OnWindowClosing(Object sender, CancelEventArgs e) {
            FormClosingEventArgs ea = (FormClosingEventArgs)e;
            if(ea.CloseReason == CloseReason.UserClosing && Window.IsMain && !isEditModelExecuting &&
                ((WinShowViewStrategyBase)Application.ShowViewStrategy).Explorers.Count == 1 &&
                XtraMessageBox.Show("You are about to exit the application. Do you want to proceed?", "Exit", MessageBoxButtons.YesNo, MessageBoxIcon.Information) == DialogResult.No) {
                e.Cancel = true;
            }
        }
        private void OnWindowClosed(object sender, EventArgs e) {
            ((WinWindow)sender).Closing -= OnWindowClosing;
            ((WinWindow)sender).Closed -= OnWindowClosed;
        }
        private void OnWindowTemplateChanged(object sender, EventArgs e) {
            Window.TemplateChanged -= OnWindowTemplateChanged;
            ((WinWindow)Window).Closing += OnWindowClosing;
            ((WinWindow)Window).Closed += OnWindowClosed;
        }
        protected override void OnDeactivated() {
            if(editModelController != null) {
                editModelController.EditModelAction.Executing -= new CancelEventHandler(EditModelAction_Executing);
                editModelController.EditModelAction.ExecuteCompleted -= new EventHandler<DevExpress.ExpressApp.Actions.ActionBaseEventArgs>(EditModelAction_ExecuteCompleted);
                editModelController = null;
            }
            Window.TemplateChanged -= OnWindowTemplateChanged;
            ((WinWindow)Window).Closing += OnWindowClosing;
            ((WinWindow)Window).Closed += OnWindowClosed;
            base.OnDeactivated();
        }
    }
}

File: VB/WinSolution.Module.Win/PromptClosingLastMainWindowController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows.Forms
Imports DevExpress.ExpressApp
Imports System.ComponentModel
Imports DevExpress.XtraEditors
Imports DevExpress.ExpressApp.Win
Imports DevExpress.ExpressApp.Win.SystemModule

Namespace WinSolution.Module.Win
	Public Class PromptClosingLastMainWindowController
		Inherits WindowController
		Private isEditModelExecuting As Boolean = False
		Private editModelController As EditModelController
		Protected Overrides Sub OnActivated()
			MyBase.OnActivated()
			editModelController = Window.GetController(Of EditModelController)()
			If editModelController IsNot Nothing Then
				AddHandler editModelController.EditModelAction.Executing, AddressOf EditModelAction_Executing
				AddHandler editModelController.EditModelAction.ExecuteCompleted, AddressOf EditModelAction_ExecuteCompleted
			End If
		End Sub
		Private Sub EditModelAction_ExecuteCompleted(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.Actions.ActionBaseEventArgs)
			isEditModelExecuting = False
		End Sub
		Private Sub EditModelAction_Executing(ByVal sender As Object, ByVal e As CancelEventArgs)
			isEditModelExecuting = True
		End Sub
		Protected Overrides Sub OnWindowChanging(ByVal window As Window)
			MyBase.OnWindowChanging(window)
			AddHandler window.TemplateChanged, AddressOf OnWindowTemplateChanged
		End Sub
		Private Sub OnWindowClosing(ByVal sender As Object, ByVal e As CancelEventArgs)
			Dim ea As FormClosingEventArgs = CType(e, FormClosingEventArgs)
			If ea.CloseReason = CloseReason.UserClosing AndAlso Window.IsMain AndAlso (Not isEditModelExecuting) AndAlso (CType(Application.ShowViewStrategy, WinShowViewStrategyBase)).Explorers.Count = 1 AndAlso XtraMessageBox.Show("You are about to exit the application. Do you want to proceed?", "Exit", MessageBoxButtons.YesNo, MessageBoxIcon.Information) = DialogResult.No Then
				e.Cancel = True
			End If
		End Sub
		Private Sub OnWindowClosed(ByVal sender As Object, ByVal e As EventArgs)
			RemoveHandler (CType(sender, WinWindow)).Closing, AddressOf OnWindowClosing
			RemoveHandler (CType(sender, WinWindow)).Closed, AddressOf OnWindowClosed
		End Sub
		Private Sub OnWindowTemplateChanged(ByVal sender As Object, ByVal e As EventArgs)
			RemoveHandler Window.TemplateChanged, AddressOf OnWindowTemplateChanged
			AddHandler (CType(Window, WinWindow)).Closing, AddressOf OnWindowClosing
			AddHandler (CType(Window, WinWindow)).Closed, AddressOf OnWindowClosed
		End Sub
		Protected Overrides Sub OnDeactivated()
			If editModelController IsNot Nothing Then
				RemoveHandler editModelController.EditModelAction.Executing, AddressOf EditModelAction_Executing
				RemoveHandler editModelController.EditModelAction.ExecuteCompleted, AddressOf EditModelAction_ExecuteCompleted
				editModelController = Nothing
			End If
			RemoveHandler Window.TemplateChanged, AddressOf OnWindowTemplateChanged
			AddHandler (CType(Window, WinWindow)).Closing, AddressOf OnWindowClosing
			AddHandler (CType(Window, WinWindow)).Closed, AddressOf OnWindowClosed
			MyBase.OnDeactivated()
		End Sub
	End Class
End Namespace